### ISSUE TITLE
Python version limit and required libraries

### ISSUE THREAD

**Issue by michael-petersen** [ilikecubesnstuff/commensurability#2]
It looks like there is some `>=Python 3.10` functionality scattered around. Starting a list here:

- `examples/using_galpy/phase_slice.py` with `pairwise` from `itertools`

Perhaps get around by requiring `Python 3.10`?

**Comment by michael-petersen** [comment]
Another dependency:

- `from tqdm import tqdm`. This one seems less strictly necessary -- perhaps make optional? Or add to requirements.
- `astropy`
- `matplotlib`? Though this is not strictly required and could be debated.
- `h5py`?

**Comment by michael-petersen** [comment]
It actually appears that `>=Python 3.11` is required, specifically to support:

- @typing.dataclass_transform(), see [documentation](https://docs.python.org/3/library/typing.html).

**Comment by ilikecubesnstuff** [comment]
- `itertools.pairwise` is used for `Backend.iter_orbit_slices`, which isn't yet used by the library - however, I have re-implemented it for now using `itertools.tee` (copying from [the itertools documentation](https://docs.python.org/3/library/itertools.html#itertools.pairwise)).
- `typing.dataclass_transform` is just a typing utility, I have removed it for Python-version backwards compatibility.
- Dependencies:
  - `astropy` is used for the `Coordinate` classes, I have added this to the package requirements.
  - Although `tqdm` isn't strictly necessary, I will add it to the requirements for now and add a note for removing it in a later commit.
  - `h5py` and `matplotlib` can be made as class-specific requirements, so using `ExtendImports` is a reasonable solution here. I think `matplotlib` should still be made a requirement since a big part of the analysis package's functionality depends on it.

I've addressed all but the last point in the latest commit.
